### PR TITLE
[WIP] AEROGEAR-2454 AuthService implementations

### PR DESCRIPTION
**Pull request by aidenkeating**
## Motivation
JIRA: https://issues.jboss.org/browse/AEROGEAR-2454

## Description

AuthService implementations.

## Progress

- [x] Android
- [x] iOS
- [x] Documentation

### CODE REVIEW

File: Auth/Auth/AbstractAuthService.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        /// Initiate an authentication flow.
+        /// </summary>
+        /// <returns>The authenticate.</returns>
+        /// <param name="authenticateOptions">Authenticate options.</param>
+        public Task<User> Authenticate(IAuthenticateOptions authenticateOptions)
+        {
+            return Authenticator.Authenticate(authenticateOptions);
+        }
+        
+        public abstract void Configure(MobileCore core, ServiceConfiguration serviceConfiguration);
```
Comment (aidenkeating) [inline Auth/Auth/AbstractAuthService.cs]: I guess this should be removed if we're just going by constructor now?

Comment (wei-lee) [reply Auth/Auth/AbstractAuthService.cs]: yeah, let's move it to the constructor

**Comment by ziccardi** [comment]
@aidenkeating : we wrote almost the same code :D I just think there is no use in passing MobileCore as parameter since it is a singleton.

File: Auth/Auth/AbstractAuthService.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:AeroGear.Mobile.Auth.AbstractAuthService"/> class.
+        /// </summary>
+        /// <param name="mobileCore">Mobile core.</param>
+        /// <param name="serviceConfig">Service config.</param>
+        public AbstractAuthService(MobileCore mobileCore = null, ServiceConfiguration serviceConfig = null)
+        {
+            MobileCore = mobileCore ?? MobileCore.Instance;
+            var serviceConfiguration = NonNull(serviceConfig ?? mobileCore.GetServiceConfiguration(Type), "serviceConfig");
```
Comment (aidenkeating) [inline Auth/Auth/AbstractAuthService.cs]: Derp, `mobileCore` should be `MobileCore`

File: Auth/Auth/AbstractAuthService.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        /// <summary>
+        /// Initiate an authentication flow.
+        /// </summary>
+        /// <returns>The authenticate.</returns>
+        /// <param name="authenticateOptions">Authenticate options.</param>
+        public Task<User> Authenticate(IAuthenticateOptions authenticateOptions)
+        {
+            return Authenticator.Authenticate(authenticateOptions);
```
Comment (aidenkeating) [inline Auth/Auth/AbstractAuthService.cs]: This and logout and everything should ensure everything is configured.

I guess the `Configure` function could go away though if we provide `AuthenticationConfig` to `InitializeService` and then provide it to the constructor. That would be simpler.

**Comment by aidenkeating** [comment]
@wei-lee I've included some dodgy logging in stuff for the iOS example app in the latest commit. It's currently very very dodgy. WIll come back to it later.

**Comment by ziccardi** [comment]
@aidenkeating @secondsun @wei-lee The proposed implementation for the DI in this PR is something I'm currently evaluating in the DI spike.
However it does not works. See my comment in [PR1](https://github.com/aidenkeating/aerogear-xamarin-sdk/pull/1)

**Comment by secondsun** [comment]
@aidenkeating @wei-lee, @ziccardi and I paired this morning for a bit and discovered that the DI works on Android but not iOS, I did more digging and we discovered that there are iOS linking configurations which are removing the dynamically loaded Assembly.  We can use class attributes on our iOS implementation classes to make the linker keep the dynamically loaded classes.

ziccardi said he would make an issue with this follow up.

**Comment by wei-lee** [comment]
@secondsun That's fine. The main problem I have is that it should be done in a separate PR, it should not be part of the work in this PR here.

**Comment by aidenkeating** [comment]
@wei-lee @secondsun If you have no objections I will move the original AuthService implementations and Example app stuff (the stuff related to our story) in to its own PR. We can then address the two topics separately in separate PRs.

**Comment by wei-lee** [comment]
@aidenkeating if it's not too much trouble, please 😄

**Comment by secondsun** [comment]
So a PR for DI, a PR for auth services and the app?  Assuming that the auth services will be built on top of the DI pr?

**Comment by wei-lee** [comment]
@secondsun I think it should be another way around: the DI pr should be on top of the auth services PR. The DI pr is nice to have, but I don't see it as a MUST. While we do need to make sure the auth services actually work.

**Comment by secondsun** [comment]
@wei-lee Again, it isn't REALLY DI.  The mechanism we have is how MobileCore looks up services, and I feel like that is a pre requiste to how services are written.  

I don't really care either way which is "first".  I just feel like doing it your way has us unwrite the DI bits, then rewrite them.  Again it doesn't actually matter once everything is written.

**Comment by secondsun** [comment]
@wei-lee Actually looking at... things I see it is easier to merge #30 and do the service injection stuff on top of it.

**Comment by secondsun** [comment]
@aidenkeating Should we close this in favor of #30 and open a new PR with the Service Injection stuff?

**Comment by secondsun** [comment]
Clsoing in favor of #30 and #31

**Comment by aidenkeating** [comment]
@secondsun Thanks for closing this, that was my plan 😀

**Pull request closed without merging**